Add Cohesive Zone Modeling large deformation capabilities

arovinelli opened the pull request:
Adding a new base `CZMMaterialBaseIncremental` and a new `CZMInterfaceKernelIncremental` to the cohesive zone framework available in the tensor mechanics module.
The new base material combines a total Lagrangian formulation and an incremental integration approach to account for large area changes and rotations oof the cohesive zone.
The new kernel impose equilibrium in the undeformed configuration and consumes material properties computed by the incremental cohesive material
The Jacobian of the kernel is exact.

Reference ref #17156

Commit:
first commit, add CZM large deformation capabilities
ref #17156

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/693644/) on 23c0f5d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17157/site)

This comment will be updated on new commits.

jiangwen84 commented:
@arovinelli  Nice work! 

Just want to confirm: this is for large opening, but not sliding, correct?

arovinelli commented:
> @arovinelli Nice work!
> 
> Just want to confirm: this is for large opening, but not sliding, correct?

@jiangwen84  thanks,
In a few words, it always assumes  initial neighbors remain neighbors. So yes large opening, large area reduction, interface rotations, but still neighbors are fixed to the initial configuration.

arovinelli commented:
@jiangwen84 and @bwspenc this is ready for review.

jiangwen84 commented:
@arovinelli I read your ANL-20/49 report and tried to better understand your formulation.

I have a general question about your equation (3.3-3.12). I am not sure why do you need to write PK-1 `T` in terms of Cauchy `t`. If your traction-separation law computes PK-1 in the local coordinate, i.e., T = T ([u]), it seems you can just properly rotate    T from local to global in addition to rigid rotation. I think that would result in much simpler derivation. Do I miss any important aspects here?

jiangwen84 commented on modules/tensor_mechanics/include/utils/DeformationGradientTools.h:
```diff
@@ -0,0 +1,39 @@
+#pragma once
```
Would it be better to call this file as "CohesiveZoneModelTools"?

arovinelli replied:
the methods defined here are generally applicable to any deformation gradient defined over a surface, the final decision is up to you

jiangwen84 replied:
I do not see any other applications that might need the deformation gradient over a surface. Let us call it "cohesiveZoneModelTools" for now. If we find more general usage, we can rename it later.

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernelIncremental.C:
```diff
@@ -0,0 +1,168 @@
+    _dPK1traction_djumpglobal(getMaterialPropertyByName<RankTwoTensor>("dPK1traction_djumpglobal")),
+    _dPK1traction_dF(getMaterialPropertyByName<RankThreeTensor>("dPK1traction_dF"))
+{
+  if (getParam<bool>("use_displaced_mesh") == true)
+    mooseError("CZMInterfaceKernelIncremental cannot be used with use_displaced_mesh = true. Large "
+               "area changes and rotations are included if the large_kinematic parameter is true "
+               "in the associated cohesive material.");
```
you may want to suppress `use_displaced_mesh `. 

params.suppressParameter<bool>("use_displaced_mesh");

jiangwen84 commented on modules/tensor_mechanics/include/materials/czm/incremental/CZMMaterialBaseIncremental.h:
```diff
@@ -0,0 +1,274 @@
+          for (unsigned int i = 0; i < 3; i++)
+            res(j, k, l) += R4(i, j, k, l) * V(i);
+        }
+    return res;
+  }
+
+  RankThreeTensor RijRjkl(const RankTwoTensor & R2, const RankThreeTensor & R3) const
+  {
+    RankThreeTensor res;
+    for (unsigned int i = 0; i < 3; i++)
+      for (unsigned int j = 0; j < 3; j++)
+        for (unsigned int k = 0; k < 3; k++)
+          for (unsigned int l = 0; l < 3; l++)
+            res(i, k, l) += R2(i, j) * R3(j, k, l);
+    return res;
+  }
+
+  RankFourTensor RijklRjm(const RankFourTensor & R4, const RankTwoTensor & R2) const
+  {
+    RankFourTensor res;
+    res.zero();
+    for (unsigned int i = 0; i < 3; i++)
+      for (unsigned int k = 0; k < 3; k++)
+        for (unsigned int l = 0; l < 3; l++)
+          for (unsigned int m = 0; m < 3; m++)
+            for (unsigned int j = 0; j < 3; j++)
+              res(i, m, k, l) += R4(i, j, k, l) * R2(j, m);
+    return res;
+  }
+
+  RankThreeTensor RjkVi(const RankTwoTensor & R2, const RealVectorValue & V) const
+  {
+    RankThreeTensor res;
+
+    for (unsigned int i = 0; i < 3; i++)
+      for (unsigned int j = 0; j < 3; j++)
+        for (unsigned int k = 0; k < 3; k++)
+          res(i, j, k) = V(i) * R2(j, k);
+    return res;
+  }
```
Could you move those functions to 'CohesiveZoneMethodTools"?

arovinelli replied:
I move this to `AdditionalTensorTools.h`

jiangwen84 commented on modules/tensor_mechanics/doc/content/syntax/Modules/TensorMechanics/CohesiveZoneMaster/index.md:
```diff
@@ -1,35 +1,51 @@
+Large deformations are supported only by the incremental formulation.
+To account for the effect of large area changes and rotations the user shall  set `large_kinematic=true` in the appropriate material block.
+
+Notice that `use_displaced_mesh = true` is not supported by any CZM material or Interfacekernel. If the user tries to manually set `use_displaced_mesh = true` in a CZM Material, InterfaceKernel or Action block an error will appear.
```
suppress 'use_displaced_mesh` param.

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/test/tests/czm/czm_large_deformation.i:
```diff
@@ -0,0 +1,355 @@
+  [./TensorMechanics]
+    [./Master]
+      [./all]
+        strain = FINITE
```
Do all those combination work?
1. bulk: small total formulation, interface: small or large
2. bulk: small incremental formulation, interface: small or large
3. bulk: finite strain, interface, small or large

arovinelli replied:
yes, all of them work. 
But you should be careful about what you do, as you are using different stresses for imposing equilibrium
The interface small formulation should be used together with a small bulk formulation (incremental or total), 
The interface large formulation should be used together with a finite strain formulation. 
you should not use the large interface model with a small bulk formulation.

jiangwen84 replied:
Could you add those information to the documentation? Thanks.

jiangwen84 reviewed (changes requested):
Overall is good. I left a few general comments. 

I have not read through all details derivation and implementation. I might have some more comments there.

jiangwen84 commented:
> @arovinelli I read your ANL-20/49 report and tried to better understand your formulation.
> 
> I have a general question about your equation (3.3-3.12). I am not sure why do you need to write PK-1 `T` in terms of Cauchy `t`. If your traction-separation law computes PK-1 in the local coordinate, i.e., T = T ([u]), it seems you can just properly rotate T from local to global in addition to rigid rotation. I think that would result in much simpler derivation. Do I miss any important aspects here?

Any thoughts?

arovinelli commented:
> > @arovinelli I read your ANL-20/49 report and tried to better understand your formulation.
> > I have a general question about your equation (3.3-3.12). I am not sure why do you need to write PK-1 `T` in terms of Cauchy `t`. If your traction-separation law computes PK-1 in the local coordinate, i.e., T = T ([u]), it seems you can just properly rotate T from local to global in addition to rigid rotation. I think that would result in much simpler derivation. Do I miss any important aspects here?
> 
> Any thoughts?

Hi wen, sorry I was replying and then I had a call. 
The short answer is we assume the traction separation law being defined in the deformed configuration while we elected to impose equilibrium in the undeformed configuration. Hence, `t` to evolve the material and `T` to impose equilibrium.

The long answer is
we decided to use an objective traction formulation because it allows us to implement stateful cohesive laws using rate formulations.
The traction separation law is defined on the interface midplane in the deformed configuration using the interface natural coordinate system. Hence you have a local  Cauchy traction `t_local`. To obtain the Cauchy traction `t` in the global reference system you have to rotate from the local to the global deformed reference system, using the rotation matrix `Q`.
The Cauchy traction on the interface midplane is different from the traction on the interface `+` and `-` surfaces. 
Notice that `t^+` and `t^-` might be different because elements on both sides might deform and rotate differently.
Imposing equilibrium using `t^+` and `t^-` would result in a more complicated formulation.
In the undeformed configuration, the `+` and `-` surfaces coincide, hence the traction must be equal and opposite. The traction in the undeformed configuration is the PK-1 traction `T`. This is why we need both.

Does this help?

jiangwen84 commented:
> > > @arovinelli I read your ANL-20/49 report and tried to better understand your formulation.
> > > I have a general question about your equation (3.3-3.12). I am not sure why do you need to write PK-1 `T` in terms of Cauchy `t`. If your traction-separation law computes PK-1 in the local coordinate, i.e., T = T ([u]), it seems you can just properly rotate T from local to global in addition to rigid rotation. I think that would result in much simpler derivation. Do I miss any important aspects here?
> > 
> > 
> > Any thoughts?
> 
> Hi wen, sorry I was replying and then I had a call.
> The short answer is we assume the traction separation law being defined in the deformed configuration while we elected to impose equilibrium in the undeformed configuration. Hence, `t` to evolve the material and `T` to impose equilibrium.
> 
> The long answer is
> we decided to use an objective traction formulation because it allows us to implement stateful cohesive laws using rate formulations.
> The traction separation law is defined on the interface midplane in the deformed configuration using the interface natural coordinate system. Hence you have a local Cauchy traction `t_local`. To obtain the Cauchy traction `t` in the global reference system you have to rotate from the local to the global deformed reference system, using the rotation matrix `Q`.
> The Cauchy traction on the interface midplane is different from the traction on the interface `+` and `-` surfaces.
> Notice that `t^+` and `t^-` might be different because elements on both sides might deform and rotate differently.
> Imposing equilibrium using `t^+` and `t^-` would result in a more complicated formulation.
> In the undeformed configuration, the `+` and `-` surfaces coincide, hence the traction must be equal and opposite. The traction in the undeformed configuration is the PK-1 traction `T`. This is why we need both.
> 
> Does this help?

I guess what I do not understand why the traction separation law has to be defined on the current configuration. Can it be defined on reference configuration such that we can evolve traction with `pk1`?

arovinelli commented:
> I guess what I do not understand why the traction separation law has to be defined on the current configuration. Can it be defined on reference configuration such that we can evolve traction with `pk1`?

Well you could assume your traction separation law lives in the undeformed configuration, but you would still need a jump vector `[u]` which is properly defined in the local deformed configuration. So it would not be kinematically consistent. And still, if you need the Cauchy traction for other reasons you would still need to remap it.

jiangwen84 commented:
> > I guess what I do not understand why the traction separation law has to be defined on the current configuration. Can it be defined on reference configuration such that we can evolve traction with `pk1`?
> 
> Well you could assume your traction separation law lives in the undeformed configuration, but you would still need a jump vector `[u]` which is properly defined in the local deformed configuration. So it would not be kinematically consistent. And still, if you need the Cauchy traction for other reasons you would still need to remap it.

Does it matter whether `[u]` is defined in the deformed or undeformed configuration? I am fine with your current approach, but I would like to know whether we could simplify it by calculating `T[u]` completely on undeformed configuration. Thanks.

arovinelli commented:
> > > I guess what I do not understand why the traction separation law has to be defined on the current configuration. Can it be defined on reference configuration such that we can evolve traction with `pk1`?
> > 
> > 
> > Well you could assume your traction separation law lives in the undeformed configuration, but you would still need a jump vector `[u]` which is properly defined in the local deformed configuration. So it would not be kinematically consistent. And still, if you need the Cauchy traction for other reasons you would still need to remap it.
> 
> Does it matter whether `[u]` is defined in the deformed or undeformed configuration? I am fine with your current approach, but I would like to know whether we could simplify it by calculating `T[u]` completely on undeformed configuration. Thanks.


Wen as I said before, technically you can. But I can't say for sure right now if that would be an easier formulation.
If you want, I can look at it, but I need some time to derive the complete formualtion.
Still, to compute the correct `[u]` we will still need all the kinematics, as it is a function of the midplane normal, which changes with deformation.

arovinelli commented:
@jiangwen84 I and @reverendbedford have been discussing your question,
now we have a question for you.
Do you want us to change where the constitutive model lives (e.g. you want to assume \dot{T}_{PK1} = f([u]) instead of  \dot{t}_{cauchy} = f([u])), or you are simply looking for a different (possibly simpler) implementation?

jiangwen84 commented:
> @jiangwen84, I and @reverendbedford have been discussing your question,
> now we have a question for you.
> Do you want us to change where the constitutive model lives (e.g. you want to assume \dot{T}_{PK1} = f([u]) instead of \dot{t}_{cauchy} = f([u])), or you are simply looking for a different (possibly simpler) implementation?

@arovinelli I think your current implementation is good to me and the formulation does not need to be changed. All I want to discuss is whether there would be a simpler implementation. We can discuss that offline for more details.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/734386/) on 66e7b33 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17157/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17157/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2da79c5b325db5af9ac088baef21cee8bd1c2cc7`

jiangwen84 commented:
@arovinelli I quickly went through your new codes. They look really great. 

If I remember correctly, in your old codes, the traction separation law in local and global coordinates is computed in a single class, but now is separated into two different classes. Is my understanding correct? 

I think you can finish your documentation and tests, I will take a close look at then.

arovinelli commented:
> @arovinelli I quickly went through your new codes. They look really great.
> 
> If I remember correctly, in your old codes, the traction separation law in local and global coordinates is computed in a single class, but now is separated into two different classes. Is my understanding correct?
> 
> I think you can finish your documentation and tests, I will take a close look at then.

Yes, now the traction separation laws are only implemented in local coordinates. They use the `the displacement jump` in local coordinates calculated by a `displacement jump provider`, which can account for large or small deformation.
Once the traction in local coordinate is calculated by the constitutive model, a `traction equilibrium calculator`, which is responsible to provide top the interface kernel the proper traction. For example, if you are using the total lagrangian formualtion the traction equilibrium calculator will compute the PK1 traction and its derivates starting from the traction and its derivatives in the local frame.
The new `cohesive zone master action` takes care of setting the correct `jump provider`, `equilibrium traction calculator` and `interface kernel` depending on the user inputs, i.e. `strain=SMALL/FINITE`, `incremental=True/False`.
In this way, you write a single material model in local coordinates and you can use it with different "strain: formulations.

jiangwen84 commented:
> > @arovinelli I quickly went through your new codes. They look really great.
> > If I remember correctly, in your old codes, the traction separation law in local and global coordinates is computed in a single class, but now is separated into two different classes. Is my understanding correct?
> > I think you can finish your documentation and tests, I will take a close look at then.
> 
> Yes, now the traction separation laws are only implemented in local coordinates. They use the `the displacement jump` in local coordinates calculated by a `displacement jump provider`, which can account for large or small deformation.
> Once the traction in local coordinate is calculated by the constitutive model, a `traction equilibrium calculator`, compute the proper traction depending on the selected formulation. For example, if you are using the total Lagrangian formulation the `traction equilibrium calculator` will compute the PK1 traction and its derivates starting from the local traction and its derivatives in the local frame.
> The new `cohesive zone master action` takes care of setting the correct `jump provider`, `equilibrium traction calculator`, and `interface kernel` depending on the user inputs, i.e. `strain=SMALL/FINITE`, `incremental=True/False`.
> In this way, you write a single material model in local coordinates and you can use it with different "strain: formulations.

I like your new design 👍

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/738043/) on 82b1983 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17157/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17157/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2da79c5b325db5af9ac088baef21cee8bd1c2cc7`

arovinelli commented:
@jiangwen84 this is ready for review

arovinelli commented:
ping @jiangwen84

jiangwen84 commented:
> ping @jiangwen84

@arovinelli Sorry for the delay. I will review it later this week.

arovinelli commented:
@jiangwen84 ping

jiangwen84 commented on modules/tensor_mechanics/doc/content/syntax/Modules/TensorMechanics/CohesiveZoneMaster/index.md:
```diff
@@ -1,35 +1,69 @@
 1. `BreakMeshByBlockGenerator`
-2. `CZMInterfaceKernels`
-3. `CZMMaterials`
+2. `DisplacementJumpProvider`
```
computeDisplacementJump?

jiangwen84 commented on modules/tensor_mechanics/doc/content/syntax/Modules/TensorMechanics/CohesiveZoneMaster/index.md:
```diff
@@ -1,35 +1,69 @@
-3. `CZMMaterials`
+2. `DisplacementJumpProvider`
+3. `CohesiveZoneConstituiveModel`
+4. `EquilibriumTractionCalcualtor`
```
computeXXXX?

jiangwen84 commented on modules/tensor_mechanics/include/materials/czm/CZMConstitutiveModelIncrementalBase.h:
```diff
@@ -0,0 +1,51 @@
+ *isotropic behavior in the tangential directions.
+ */
+
+class CZMConstitutiveModelIncrementalBase : public CZMConstitutiveModelTotalBase
```
It is a bit wired to derive `Incremental` from `Total`. Can they derive from a same base class?

jiangwen84 commented on modules/tensor_mechanics/include/materials/czm/CZMConstitutiveModelTotalBase.h:
```diff
@@ -0,0 +1,43 @@
+ * orthogonal tangential components. The model also assumes isotropic behavior in the tangential
+ * directions.
+ */
+class CZMConstitutiveModelTotalBase : public InterfaceMaterial
```
`CZMConstitutiveModelxxx` is used to compute the traction in the local interface coordinates? If so, I would like to rename it to `computeCZMLocalTraction` or something like that.

jiangwen84 commented on modules/tensor_mechanics/include/materials/czm/CZMEquilibriumTractionCalculatorBase.h:
```diff
@@ -0,0 +1,44 @@
+ * the global displacement jump, starting from the values provided from any CZM constituive material
+ * model.
+ */
+class CZMEquilibriumTractionCalculatorBase : public InterfaceMaterial
```
As I mentioned before, I would like to rename it `computeCZMGlobalTraction` or something similar. I just want to let users easily understand those are in local or global systems.

jiangwen84 commented on modules/tensor_mechanics/include/materials/czm/SalehaniIrani3DCTractionIncremental.h:
```diff
@@ -0,0 +1,36 @@
+//* This file is part of the MOOSE framework
```
Can we rename the folder name 'czm' to 'cohesive_zone_model`?

jiangwen84 reviewed (changes requested):
In my first round review, I would like to get some high-level suggestions.

Commit:
rename files and directory

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/753475/) on 239cb36 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17157/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17157/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format dfb54e37fe0f13a4a77248f1aa4d7d7f2e9cd7e6`

Commit:
style patch

arovinelli commented:
@jiangwen84 i updated the naming as suggested. Please check and make sure you like them. I also added a base class for the total and incremental material models. If naming is fine please go ahead with a second review.

Commit:
add missing mesh file

moosebuild commented:
Job [Test Heavy](https://civet.inl.gov/job/753880/) on d03513b : invalidated by @jiangwen84

arovinelli commented:
@jiangwen84 do you like the new names?

jiangwen84 commented on modules/tensor_mechanics/src/actions/CohesiveZoneMasterAction.C:
```diff
@@ -45,19 +64,43 @@ CohesiveZoneMasterAction::act()
+  else if (_current_task == "add_material")
+  {
+    // Create unique material name for the displacement jump provider
+    std::string unique_material_name = _disp_jump_provider_name + "_" + _name;
+    InputParameters paramsm = _factory.getValidParams(_disp_jump_provider_name);
+    paramsm.set<std::vector<BoundaryName>>("boundary") =
+        getParam<std::vector<BoundaryName>>("boundary");
+    paramsm.set<std::vector<VariableName>>("displacements") = displacements;
+    _problem->addInterfaceMaterial(_disp_jump_provider_name, unique_material_name, paramsm);
```
Can we set different traction-separation law for different boundaries using the Action?

arovinelli replied:
need top try, I'll make a test for this

arovinelli replied:
The answer is yes.
I added a test for this case

jiangwen84 commented on modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernelSmallStrain.C:
```diff
@@ -108,9 +106,9 @@ CZMInterfaceKernel::computeQpOffDiagJacobian(Moose::DGJacobianType type, unsigne
       break;
```
The logic here is not quite right. Either I fix it in a separate PR or you fix it here.

You need to consider there might be other variables in the problem.

arovinelli replied:
I think I fixed it

jiangwen84 commented on modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernelTotalLagrangian.C:
```diff
@@ -0,0 +1,164 @@
+      jac += _test_neighbor[_i][_qp] * JacLD(_component, /*neighbor=*/true);
+      break;
+  }
+  mooseAssert(std::isfinite(jac), "CZMInterfaceKernelTotalLagrangian diag jacobian is not finite");
```
why do we need this check?

arovinelli replied:
this is a leftover of my development I'll remove it

jiangwen84 commented on modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernelTotalLagrangian.C:
```diff
@@ -0,0 +1,164 @@
+{
+
+  // find the displacement component associated to jvar
+  unsigned int off_diag_component;
+  for (off_diag_component = 0; off_diag_component < _ndisp; off_diag_component++)
+    if (_disp_var[off_diag_component] == jvar)
+      break;
```
Change the logic. You can follow what was done in `StressDivergenceTensors.C`

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/CZMComputeDisplacementJumpSmallStrain.C:
```diff
@@ -0,0 +1,43 @@
+}
+
+void
+CZMComputeDisplacementJumpSmallStrain::initQpStatefulProperties()
+{
+  // requried to promote _interface_displacement_jump to stateful in case someone needs it
+  _interface_displacement_jump[_qp] = 0;
+}
```
Can we move `_interface_displacement_jump` to the base class?

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/CZMComputeDisplacementJumpTotalLagrangian.C:
```diff
@@ -0,0 +1,117 @@
+
+CZMComputeDisplacementJumpTotalLagrangian::CZMComputeDisplacementJumpTotalLagrangian(
+    const InputParameters & parameters)
+  : CZMComputeDisplacementJumpSmallStrain(parameters),
```
I thought we decided to derive `CZMComputeDisplacementJumpTotalLagrangian` and `CZMComputeDisplacementJumpSmallStrain` from a same base class. Correct?

arovinelli replied:
ups, I forgot about this. 
As requested I added a base class

jiangwen84 reviewed (changes requested):
I finished my second round review. A few minor comments.

Commit:
almost ready

Commit:
resolve comments

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/770490/) on 4505592 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17157/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17157/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d4a83b1aa02b896d00b6ef07dcc82efeedf38b52`

Commit:
patch

arovinelli commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/CZMComputeGlobalTractionSmallStrain.h:
```diff
@@ -0,0 +1,29 @@
+
+protected:
+  /// method used to compute the traction and it's derivatives in global coordinates.
+  void computeEquilibriumTracionAndDerivatives() override final;
```
removed final

arovinelli commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/CZMComputeGlobalTractionTotalLagrangian.h:
```diff
@@ -0,0 +1,108 @@
+
+  MaterialProperty<RankThreeTensor> & _dPK1traction_dF;
+
+  /// the area raio and the area increment rate
```
done

arovinelli commented on modules/tensor_mechanics/include/interfacekernels/CZMInterfaceKernelSmallStrain.h:
```diff
@@ -15,11 +15,11 @@
 /// a single displacement compenent.
 /// One kernel is required for each mesh dimension.
-class CZMInterfaceKernel : public InterfaceKernel
+class CZMInterfaceKernelSmallStrain : public InterfaceKernel
```
added `base_name` to all objects. There is also a test usign them

arovinelli commented on modules/tensor_mechanics/src/actions/CohesiveZoneMasterAction.C:
```diff
@@ -25,17 +27,34 @@ CohesiveZoneMasterAction::validParams()
+    _equilibrium_traction_calculator_name = "CZMComputeGlobalTractionTotalLagrangian";
+  }
+  else
+    mooseError("Internal error");
```
done

arovinelli commented on modules/tensor_mechanics/include/utils/AdditionalTensorTools.h:
```diff
@@ -0,0 +1,33 @@
+#include "RankTwoTensor.h"
+#include "Assembly.h"
+
+namespace AdditionalTensorTools
```
fixed by #18174

arovinelli commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/SalehaniIrani3DCTraction.h:
```diff
@@ -0,0 +1,40 @@
+  const std::vector<Real> _delta_u0;
+
+  /// the vector representing the maximum allowed traction in each direction
+  const std::vector<Real> _max_allowable_traction;
```
done

arovinelli commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/CZMComputeDisplacementJumpTotalLagrangian.h:
```diff
@@ -0,0 +1,69 @@
+
+  /// compute the interface displacement increment using an incremental
+  /// total Lagrangian approach
+  void computeLocalDisplacementJump() override final;
```
removed the `final` attribute

arovinelli commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/CZMComputeDisplacementJumpBase.h:
```diff
@@ -0,0 +1,52 @@
+  ///@}
+
+  /// the rotation matrix transforming from the interface to the global coordinate systems in the undeformed configuration
+  MaterialProperty<RealTensorValue> & _Q0;
```
renamed this to `czm_reference_rotation`

arovinelli commented on modules/tensor_mechanics/include/interfacekernels/CZMInterfaceKernelTotalLagrangian.h:
```diff
@@ -0,0 +1,54 @@
+/// DG kernel implementing cohesive zone models (CZM) for a 1D/2D/3D traction
+/// separation laws based on the displacement jump. This kernel operates only on
+/// a single displacement compenent.
+/// One kernel is required for each mesh dimension.
```
done

arovinelli commented on modules/tensor_mechanics/include/interfacekernels/CZMInterfaceKernelSmallStrain.h:
```diff
@@ -15,11 +15,11 @@
 /// separation laws based on the displacement jump. This kernel operates only on
 /// a single displacement compenent.
 /// One kernel is required for each mesh dimension.
```
done

arovinelli commented on modules/tensor_mechanics/doc/content/syntax/Modules/TensorMechanics/CohesiveZoneMaster/index.md:
```diff
@@ -1,35 +1,69 @@
-3. `CZMMaterials`
+2. `DisplacementJumpProvider`
+3. `CohesiveZoneConstituiveModel`
+4. `EquilibriumTractionCalcualtor`
```
done

Commit:
fix test file

Commit:
another fix

Commit:
resolve comments and improve documentation

jiangwen84 commented on modules/tensor_mechanics/include/actions/CohesiveZoneMasterAction.h:
```diff
@@ -17,5 +17,33 @@ class CohesiveZoneMasterAction : public Action
+  /// strain formulation
+  enum class Kinematic
+  {
+    SmallStrain,
+    TotalLagrangian
```
Shouldn't it be `SmallStrain` v.s. `FiniteStrain` or `Incremental` v.s. `Total`?

arovinelli replied:
This is really up to you. The input parameter name is `kinematic`, not `strain` so as it is now it make sense, at least for me this nomenclature.
If you prefer strain, let me know but then, in case someone wants to add an updated lagrangian formulation, we will need a `kinematic` parameter distinguishing between the two formulations. Just let me know

arovinelli replied:
@jiangwen84 can you tell me your thoughts about this? should I change the parameter `kinematic`->`strain` and use `Finite Strain`

jiangwen84 replied:
To make it consistent with TM, I prefer use Small and Finite as options. What do you think?

arovinelli replied:
Fair enough, I'll make the change

jiangwen84 commented on modules/tensor_mechanics/include/interfacekernels/CZMInterfaceKernelTotalLagrangian.h:
```diff
@@ -0,0 +1,33 @@
+#include "CZMInterfaceKernelBase.h"
+
+/// DG cohesive zone model kernel for the Total Lagrangian formulation.
+/// This kernel assummes teh traction sepration law only depends from the
```
teh -> the

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/CZMComputeDisplacementJumpTotalLagrangian.h:
```diff
@@ -0,0 +1,69 @@
+
+#include "CZMComputeDisplacementJumpBase.h"
+/**
+ * Compute the displacement jump in interface coordinates accross a cohesive zone for the total
```
accross -> across

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/CZMComputeLocalTractionIncremental.h:
```diff
@@ -0,0 +1,49 @@
+ *isotropic behavior in the tangential directions.
+ */
+
+class CZMComputeLocalTractionIncremental : public CZMComputeLocalTractionBase
```
Since this is also a base class, shall we rename it to CZMComputeLocalTractionIncrementalBase?

arovinelli replied:
ok, agree!

arovinelli replied:
agree

jiangwen84 commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/CZMComputeLocalTractionTotal.h:
```diff
@@ -0,0 +1,28 @@
+ * orthogonal tangential components. The model also assumes isotropic behavior in the tangential
+ * directions.
+ */
+class CZMComputeLocalTractionTotal : public CZMComputeLocalTractionBase
```
Same here, add the `base` at the end?

arovinelli replied:
as above, agree

arovinelli replied:
agree

jiangwen84 commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/CZMComputeLocalTractionTotal.h:
```diff
@@ -0,0 +1,28 @@
+
+#include "CZMComputeLocalTractionBase.h"
+/**
+ * Base class used to implement traction separetion laws. All traction separetion laws
+ * shall compute the interface traction using the interface coordinate system, and
+ * traction derivtives w.r.t. to the interface displacement jump. Interface traction and related
+ * derivatives should be implemented overriding the computeInterfaceTractionAndDerivatives method.
+ * The interface coordinate system assumes the three component of the traction and
+ * disaplcement jump being ordered as [N,S1,S2], where N is the normal component and S1, S2 two
+ * orthogonal tangential components. The model also assumes isotropic behavior in the tangential
+ * directions.
```
Highlight this is for total formulation. Otherwise, the comment is almost the same as the `incremental` one.

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/SalehaniIrani3DCTraction.h:
```diff
@@ -0,0 +1,40 @@
+ * Implementation of the non-stateful exponential traction separation law
+ * proposed by Salehani, Mohsen Khajeh and Irani, Nilgoon 2018
+ **/
+class SalehaniIrani3DCTraction : public CZMComputeLocalTractionTotal
```
Since this traction separation law works for both 2D and 3D, maybe we should remove `3D` from its name. User might think it is 3D only.

arovinelli replied:
All traction separation laws can be used in 1D, 2D and 3D, but hey always must eb coded for the 3D case. In this specific case, 3DC is part of the model name, as it is an extension of the Xu-Needleman model to the 3D case

jiangwen84 replied:
what does `3DC` stand for?

arovinelli replied:
3DC stands for 3D contact, and is the name the authors gave to the model.  Actually I was wrong about the origin, this model is an extension of the McGarry et al.  So I have no particular preference, which itself is an extension of the model from Bosch et al.

jiangwen84 replied:
If authors gave the name to the model and called it in their paper, I guess 3DC is fine to use. No change is needed.

arovinelli replied:
ok, it wills tay as is

jiangwen84 commented on modules/tensor_mechanics/include/utils/CohesiveZoneModelTools.h:
```diff
@@ -0,0 +1,48 @@
+
+#include "RankFourTensor.h"
+#include "RankTwoTensor.h"
+#include "Assembly.h"
```
Do you really need `Assembly.h`?

arovinelli replied:
need to check why this was here

arovinelli replied:
we only need `#include libmesh/vector_value.h`

jiangwen84 commented on modules/tensor_mechanics/src/actions/CohesiveZoneMasterAction.C:
```diff
@@ -23,41 +25,100 @@ CohesiveZoneMasterAction::validParams()
+                   : ""),
+    _kinematic(getParam<MooseEnum>("kinematic").getEnum<Kinematic>())
 {
+  // We can't enforce conssintency between the number of displacement varaiables and the mesh
```
consistency

jiangwen84 commented on modules/tensor_mechanics/src/actions/CohesiveZoneMasterAction.C:
```diff
@@ -23,41 +25,100 @@ CohesiveZoneMasterAction::validParams()
+  // We can't enforce conssintency between the number of displacement varaiables and the mesh
+  // dimension. Hence we only check we have a reasonable number of displacement variables
+  if (_ndisp > 3 || _ndisp < 1)
+    mooseError("the CZM Action requires 1, 2 or 3 displacement variables.");
```
I am curious whether TM action checks this.

arovinelli replied:
If I remember correctly it does not. Is there any reason why you don't enforce this?

jiangwen84 commented on modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernelBase.C:
```diff
@@ -0,0 +1,108 @@
+                                        " component == 0, ==> X"
+                                        " component == 1, ==> Y"
+                                        " component == 2, ==> Z");
+  params.set<bool>("_use_undisplaced_reference_points") = true;
```
what is this `_use_undisplaced_reference_points`?

arovinelli replied:
this is a parameter allowing to use fake neighbors when a kernel is used with 'displaced_mesh=True'. At the moment all CZMinterface kernel use 'displaced_mesh=false', but it should stay there until fake neighbors are not part of libmesh

jiangwen84 commented on modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernelBase.C:
```diff
@@ -0,0 +1,108 @@
+}
+
+void
+CZMInterfaceKernelBase::initialSetup()
+{
+  for (unsigned int i = 0; i < _ndisp; ++i)
+  {
+    _disp_var[i] = coupled("displacements", i);
+    _disp_neighbor_var[i] = coupled("displacements", i);
+    _vars[i] = getVar("displacements", i);
+  }
+}
```
why not in the constructor?

arovinelli replied:
At construction time the constant reference might not have been set yet. Also this appears to be a cleaner implementation, though it would require explicitly calling it in derived classes.

jiangwen84 commented on modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernelBase.C:
```diff
@@ -0,0 +1,108 @@
+Real
+CZMInterfaceKernelBase::computeQpOffDiagJacobian(Moose::DGJacobianType type, unsigned int jvar)
+{
+  Real jac = 0.0;
```
do you need `jac` ?

arovinelli replied:
removed the unnecessary variable

jiangwen84 commented on modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernelBase.C:
```diff
@@ -0,0 +1,108 @@
+{
+  Real jac = 0.0;
+  // bail out if jvar is not coupled
+  if (getJvarMap()[jvar] < 0)
+    return jac;
```
I am not familiar with this function. Do we really need this?

arovinelli replied:
this function catches if a variable is coupled with the kernel or not. If it is not it returns a negative value. This is how the `StressDiveregenceKernel` works, as per you previous comment

jiangwen84 commented on modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernelTotalLagrangian.C:
```diff
@@ -0,0 +1,74 @@
+  for (unsigned int j = 0; j < 3; j++)
+    jacld += _dPK1traction_dF[_qp](_component, cc, j) * phi(j);
+
+  mooseAssert(std::isfinite(jacld), "CZMInterfaceKernelTotalLagrangian JacLD is not finite");
```
I am curious how useful this assertion will be.

arovinelli replied:
this allows to catch problem in debug mode when developing new materials. It's an extra layer of security and it doesn't impact performance (assertions are only evaluated in debug mode)

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/CZMComputeDisplacementJumpBase.C:
```diff
@@ -0,0 +1,90 @@
+
+void
+CZMComputeDisplacementJumpBase::initialSetup()
+{
+  // initializing the displacement vectors
+  for (unsigned int i = 0; i < _ndisp; ++i)
+  {
+    _disp[i] = &coupledValue("displacements", i);
+    _disp_neighbor[i] = &coupledNeighborValue("displacements", i);
+  }
+
+  // All others zero (so this will work naturally for 2D and 1D problems)
+  for (unsigned int i = _ndisp; i < 3; i++)
+  {
+    _disp[i] = &_zero;
+    _disp_neighbor[i] = &_zero;
+  }
+}
```
same here. why this not in the constructor?

arovinelli replied:
see above

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/CZMComputeDisplacementJumpBase.C:
```diff
@@ -0,0 +1,90 @@
+void
+CZMComputeDisplacementJumpBase::computeQpProperties()
+{
+
```
remove the empty line.

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/CZMComputeDisplacementJumpTotalLagrangian.C:
```diff
@@ -0,0 +1,114 @@
+  // unwanted overflow (I tried and it happens). So checking component by
+  // component remains the only reasonable strategy. If someone finds a  better
+  // way this could be changed in the future
+  for (uint i = 0; i < 3; i++)
+    for (uint j = 0; j < 3; j++)
```
unit -> unsigned int

arovinelli replied:
ok

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/CZMComputeDisplacementJumpTotalLagrangian.C:
```diff
@@ -0,0 +1,114 @@
+  // way this could be changed in the future
+  for (uint i = 0; i < 3; i++)
+    for (uint j = 0; j < 3; j++)
+      if (!std::isfinite(_F[_qp](i, j)))
+        throw MooseException("CZMMaterialBaseIncremental _F is not finite, reducing time step");
```
What is the consequence of F having infinite value? Would solver terminate? Or it can automatically cut time step?

arovinelli replied:
nonfinite values trigger termination, so this is a safeguard to reduce tilmestep before making the all simulation crash. I have't found any other way around it, and in CP simulations sometimes it happens

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/CZMComputeGlobalTractionTotalLagrangian.C:
```diff
@@ -0,0 +1,138 @@
+void
+CZMComputeGlobalTractionTotalLagrangian::computedTPK1dJumpGlobal()
+{
+
```
remove the empty line.

arovinelli replied:
done

jiangwen84 reviewed (changes requested):
A few more minor comments.

Commit:
resolve second round of comments and adds save_in methods to the action

arovinelli commented:
@jiangwen84 everything should be resolved. At this point the only thing missing comparing to the TM action are the automatic output generation. Please recheck and let me know if you want me to add the outputs like `traction_global_x/y/z` and `displacement_jump_x/y/z` etc

jiangwen84 commented:
> @jiangwen84 everything should be resolved. At this point the only thing missing comparing to the TM action are the automatic output generation. Please recheck and let me know if you want me to add the outputs like `traction_global_x/y/z` and `displacement_jump_x/y/z` etc

If it is not much work, could you add them? I think they will be very useful. BTW, how do you visualize them in paraview? They only live on lower-dimension element, correct?

arovinelli commented:
> > @jiangwen84 everything should be resolved. At this point the only thing missing comparing to the TM action are the automatic output generation. Please recheck and let me know if you want me to add the outputs like `traction_global_x/y/z` and `displacement_jump_x/y/z` etc
> 
> If it is not much work, could you add them? I think they will be very useful. BTW, how do you visualize them in paraview? They only live on lower-dimension element, correct?

at the moment, the only way to visualize them is to setup a boundary restricted aux kernel and you will get values inside the cell... this is an approximation because if you have more then one side belonging to a boundary you will get an average value. They should live on an LD element, but I don't know how to transfer them from solid elements to LD elements. This task is part of the NEAMS technical help `Output boundary and interface material properties and aux variables on surfaces`, but still no progress on it. Back to your question, in paraview you can use a `cell_to_node` filter and then  plot the glyph of the vector field generated by it. it is still better than nothing

jiangwen84 commented:
> > > @jiangwen84 everything should be resolved. At this point the only thing missing comparing to the TM action are the automatic output generation. Please recheck and let me know if you want me to add the outputs like `traction_global_x/y/z` and `displacement_jump_x/y/z` etc
> > 
> > 
> > If it is not much work, could you add them? I think they will be very useful. BTW, how do you visualize them in paraview? They only live on lower-dimension element, correct?
> 
> at the moment, the only way to visualize them is to setup a boundary restricted aux kernel and you will get values inside the cell... this is an approximation because if you have more then one side belonging to a boundary you will get an average value. They should live on an LD element, but I don't know how to transfer them from solid elements to LD elements. This task is part of the NEAMS technical help `Output boundary and interface material properties and aux variables on surfaces`, but still no progress on it. Back to your question, in paraview you can use a `cell_to_node` filter and then plot the glyph of the vector field generated by it. it is still better than nothing

Got it.I suggest we add it. Thanks!

arovinelli commented:
> > > > @jiangwen84 everything should be resolved. At this point the only thing missing comparing to the TM action are the automatic output generation. Please recheck and let me know if you want me to add the outputs like `traction_global_x/y/z` and `displacement_jump_x/y/z` etc
> > > 
> > > 
> > > If it is not much work, could you add them? I think they will be very useful. BTW, how do you visualize them in paraview? They only live on lower-dimension element, correct?
> > 
> > 
> > at the moment, the only way to visualize them is to setup a boundary restricted aux kernel and you will get values inside the cell... this is an approximation because if you have more then one side belonging to a boundary you will get an average value. They should live on an LD element, but I don't know how to transfer them from solid elements to LD elements. This task is part of the NEAMS technical help `Output boundary and interface material properties and aux variables on surfaces`, but still no progress on it. Back to your question, in paraview you can use a `cell_to_node` filter and then plot the glyph of the vector field generated by it. it is still better than nothing
> 
> Got it.I suggest we add it. Thanks!

on it, than we will update to the method when #15920 is resolved

Commit:
see what is failing

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/778462/) on c72642b : invalidated by @jiangwen84

moosebuild commented:
Job [Moose Install Target Tests](https://civet.inl.gov/job/778470/) on c72642b : invalidated by @jiangwen84

moosebuild commented:
Job [External App tests](https://civet.inl.gov/job/778461/) on c72642b : invalidated by @jiangwen84

arovinelli commented:
@jiangwen84 this is ready for you.

jiangwen84 commented on modules/tensor_mechanics/include/actions/CohesiveZoneAction.h:
```diff
@@ -0,0 +1,137 @@
+  void actOutputGeneration();
+  void actOutputMatProp();
+
+  /// adds the required interfacekernels based on teh selected strain formulation
```
teh -> the

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/include/actions/CohesiveZoneAction.h:
```diff
@@ -0,0 +1,137 @@
+  /// adds the required interfacekernels based on teh selected strain formulation
+  void addRequiredCZMInterfaceKernels();
+
+  /// adds the required interface materials based on teh selected strain formulation
```
teh -> the

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/include/actions/CohesiveZoneAction.h:
```diff
@@ -0,0 +1,137 @@
+  /// adds the required interface materials based on teh selected strain formulation
+  void addRequiredCZMInterfaceMaterials();
+
+  /// verifies oreder of family of output variables
```
order  and family

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/doc/content/syntax/Modules/TensorMechanics/CohesiveZoneMaster/index.md:
```diff
@@ -1,35 +1,90 @@
+!listing modules/tensor_mechanics/test/tests/cohesive_zone_model/stretch_rotate_large_deformation.i block=Modules/TensorMechanics/CohesiveZoneMaster
+
+If necessary, multiple instances of the `CohesiveZoneMaster` action can be added, for instance when different material properties `base_name` are needed for different boundaries. The `base_name` parameter used in the action should also be provided to the associated materials.
+
```
Could you add  description of`generate_output` parameter here?

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/doc/content/source/materials/cohesive_zone_model/CZMRealVectorCartesianComponent.md:
```diff
@@ -0,0 +1,26 @@
+This Material model is set up by [CohesiveZoneMaster](CohesiveZoneMaster/index.md) automatically
+when traction or displacement-jump components are requested in the generate_output parameter, but can also be set up directly by the user.  
+
+The `CZMRealVectorCartesianComponent` takes as arguments the values of the `index` to save into a MaterialProperty.  [eq:real_vecto_indices] shows the component values for each vector component.
```
Do we really need this? We can use `MaterialRealVectorValueAux` to fetch the material properties.

arovinelli replied:
this is also how the TM action works: first, it generates all scalar material properties, then it adds MaterialRealAux. I simply followed it, and I asked myself the same question about `RankTwoAux`. Let me know

jiangwen84 reviewed (changes requested):
A few minor comments.

arovinelli commented on modules/tensor_mechanics/doc/content/source/materials/cohesive_zone_model/PureElasticTractionSeparation.md:
```diff
@@ -0,0 +1,20 @@
+
+!syntax description /Materials/PureElasticTractionSeparation
+
+## Description
```
It provides a pure elastic traction separation law, I mostly use it for testing, but most software has it for pedagogical purposes. If you want I can move it tests.

jiangwen84 replied:
You can keep it.

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/CZMRealVectorCartesianComponent.C:
```diff
@@ -0,0 +1,50 @@
+}
+
+void
+CZMRealVectorCartesianComponent::initQpStatefulProperties()
+{
+  _property[_qp] = 0.0;
+}
```
If _property is not stateful, we do not need this.

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/CZMRealVectorScalar.C:
```diff
@@ -0,0 +1,69 @@
+      "property_name", "Name of the material property computed by this model");
+  MooseEnum scalarType("Normal Tangent");
+  params.addRequiredParam<MooseEnum>(
+      "scalar_type", scalarType, "Type of scalar to be computed: Normal, Tangent");
```
we probably need a better naming. maybe `direction_type` or something else.

There will be two tangential directions and how do you want to output them separately.

arovinelli replied:
for the tangent direction, we only output the norm of the tangent vector. Knowing the tangent components in interface coordiante is uselles, because the two tangent directions vary elemnt by element.

arovinelli replied:
I will change this to `direction`

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/CZMRealVectorScalar.C:
```diff
@@ -0,0 +1,69 @@
+    _property(declarePropertyByName<Real>(getParam<MaterialPropertyName>("property_name"))),
+    _vector(getMaterialPropertyByName<RealVectorValue>(_base_name +
+                                                       getParam<std::string>("real_vector_value"))),
+    _czm_rotation(getMaterialPropertyByName<RankTwoTensor>(_base_name + "czm_total_rotation")),
+    _unit_vector({1, 0, 0})
```
maybe we can combine the two.

_normal(getMaterialPropertyByName<RankTwoTensor>(_base_name + "czm_total_rotation")*RealVectorValue(1.0,0.0,0.0))

arovinelli replied:
I've never seen this kind of operation, I don't even think it is possible to do this in the constructor, or is it?

jiangwen84 replied:
Have you tried it?

arovinelli replied:
no but this will not work, because for large strain the `czm_total_rotation ` changes at each evaluation, while what you are suggesting is static number

jiangwen84 replied:
Ok. Then do not need `_unit_vector` as a member variable.

arovinelli replied:
done

jiangwen84 commented on modules/tensor_mechanics/src/actions/CohesiveZoneActionBase.C:
```diff
@@ -0,0 +1,152 @@
+#include "MooseEnum.h"
+
+// map vector name shortcuts to tensor material property names
+const std::map<std::string, std::string>
+    CohesiveZoneActionBase::_real_vector_cartesian_component_table = {
+        {"traction", "traction_global"},
+        {"jump", "displacement_jump_global"},
+        {"interface_jump", "interface_displacement_jump"},
+        {"pk1_traction", "PK1traction"}};
```
Can we provide the meaning of those output names in the Action Mark Down file?

arovinelli replied:
done

Commit:
address comments

arovinelli commented:
@jiangwen84 I addressed all comments, ready for you again

jiangwen84 reviewed (dismissed):
@arovinelli I am going to merge your PR once it passes all tests. Thanks again for this great work. Look forward to using this new capability.

arovinelli commented:
> @arovinelli I am going to merge your PR once it passes all tests. Thanks again for this great work. Look forward to using this new capability.

I rebased because of an unrelated error in `Mac Test`
